### PR TITLE
Refactoring/filter route

### PR DESCRIPTION
**Pull request by gmakharat**
**Thing have done:**

- Introduced `Filterable` interface to have abstraction of filter.
- Created `FilterFactory` class with its factory method to get specific filter object. (Not the best implementation since the whole logic of filtering is not well organized from the very beginning of creation of filtering).
- Created `BaseFilter` class to follow generalization, which gathers common code for both `participantList` and `filterList` parents.
- `BaseFilterParticipantList` which extends `BaseFilter` for `participantList` parent specifics.
- `BaseFilterTissueList` which extends `BaseFilter` for `tissueList` parent specifics.
- Created `GetParticipantRoute` to get `ParticipantWrapper` object by participant id.

P.S This was attempt to apply object oriented design principles to `FilterRoute` to have cleaner and understandable code, filtering design still needs to be improved in the future.

### CODE REVIEW

File: src/main/java/org/broadinstitute/dsm/model/filter/FilterFactory.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                        } else if (StringUtils.isNotBlank(queryParams.get(RequestParameter.FILTERS).value())){
+                            filterable = new SavedFilterParticipantList();
+                        } else {
+                            filterable = new EmptyFilterParticipantList();
+                        }
+                        break;
+                    case BaseFilter.TISSUE_LIST_PARENT:
+                        String filterQuery = queryParams.get("filterQuery").value();
+                        if (StringUtils.isNotBlank(filterQuery)) {
+                            filterable = new SearchBarFilterTissueList();
```
Comment (SimoneMaiwald) [inline src/main/java/org/broadinstitute/dsm/model/filter/FilterFactory.java:47]: Where do we filter the participant list search bar ?

File: src/main/java/org/broadinstitute/dsm/model/filter/participant/BaseFilterParticipantList.java [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+                    if (filter.getParticipantColumn() != null && PARTICIPANT_DATA.equals(filter.getParticipantColumn().tableAlias)) {
+                        if (allParticipantData == null) {
+                            allParticipantData = participantDataDao
+                                    .getParticipantDataByInstanceId(Integer.parseInt(instance.getDdpInstanceId()));
+                        }
+                        numberOfParticipantDataFilters++;
+                        addParticipantDataIdsForFilters(filter, tmpName, allParticipantData, allIdsForParticipantDataFiltering);
+                    } else {
+                        if (StringUtils.isNotBlank(tmpName)) {
+                            dbElement = columnNameMap.get(tmp + "." + tmpName);
```
Comment (SimoneMaiwald) [inline src/main/java/org/broadinstitute/dsm/model/filter/participant/BaseFilterParticipantList.java:81]: Could tmp be null?

File: src/main/java/org/broadinstitute/dsm/model/filter/tissue/BaseFilterTissueList.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+
+    protected List<TissueListWrapper> filterTissueList(Filter[] filters, Map<String, DBElement> columnNameMap, String filterName,
+                                                     @NonNull DDPInstance instance, String filterQuery) {
+        Map<String, String> queryConditions = new HashMap<>();
+        String subQueryForFiltering = "";
+        if (filters != null && !columnNameMap.isEmpty()) {
+            for (Filter filter : filters) {
+                if (filter != null) {
+                    DBElement dbElement = columnNameMap.get(filter.getParticipantColumn().tableAlias + "." + filter.getFilter1().getName());
```
Comment (SimoneMaiwald) [inline src/main/java/org/broadinstitute/dsm/model/filter/tissue/BaseFilterTissueList.java:40]: What if filter.getFilter1() is null?

File: src/main/java/org/broadinstitute/dsm/model/filter/tissue/BaseFilterTissueList.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                    ViewFilter.addQueryCondition(queryConditions, dbElement, filter);
+                }
+            }
+        }
+
+        if (StringUtils.isNotBlank(filterQuery) && StringUtils.isBlank(subQueryForFiltering)) {
```
Comment (SimoneMaiwald) [inline src/main/java/org/broadinstitute/dsm/model/filter/tissue/BaseFilterTissueList.java:46]: Is subQueryForFiltering not always blank?

File: src/main/java/org/broadinstitute/dsm/model/filter/tissue/BaseFilterTissueList.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            logger.info("Found query conditions for " + queryConditions.size() + " tables");
+        }
+        String queryString = "";
+        for (String key : queryConditions.keySet()) {
+            if (!"ES".equals(key)) {
+                queryString += queryConditions.get(key);
+            }
+        }
+        queryString += subQueryForFiltering;
+        return getListBasedOnFilterName(filterName, instance.getName(), "tissueList", queryString, queryConditions);
```
Comment (SimoneMaiwald) [inline src/main/java/org/broadinstitute/dsm/model/filter/tissue/BaseFilterTissueList.java]: can we use the string from BaseFilter for TISSUE_LIST_PARENT here?

File: src/main/java/org/broadinstitute/dsm/model/filter/tissue/ManualFilterTissueList.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        if (Objects.isNull(queryParamsMap)) return wrapperList;
+        prepareNeccesaryData(queryParamsMap);
+        if (!queryParamsMap.hasKey(RoutePath.REALM)) throw new NoSuchElementException("realm is necessary");
+        String realm = queryParamsMap.get(RoutePath.REALM).value();
+        DDPInstance ddpInstance = DDPInstance.getDDPInstance(realm);
+        String defaultFilter = queryParamsMap.get(RoutePath.FILTER_DEFAULT).value();
+        List<TissueList> tissueListList;
+        if ("0".equals(defaultFilter)) {
+            tissueListList = TissueList.getAllTissueListsForRealmNoFilter(realm);
+            wrapperList = TissueListWrapper.getTissueListData(ddpInstance, null, tissueListList);
```
Comment (SimoneMaiwald) [inline src/main/java/org/broadinstitute/dsm/model/filter/tissue/ManualFilterTissueList.java:41]: can we use the string from BaseFilter for TISSUE_LIST_PARENT here?

File: src/main/java/org/broadinstitute/dsm/model/filter/tissue/ManualFilterTissueList.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        } else if ("1".equals(defaultFilter)) {
+            String userEmail = queryParamsMap.value(RequestParameter.USER_MAIL);
+            String defaultFilterName = ViewFilter.getDefaultFilterForUser(userEmail, "tissueList");
```
Comment (SimoneMaiwald) [inline src/main/java/org/broadinstitute/dsm/model/filter/tissue/ManualFilterTissueList.java]: can we use the string from BaseFilter for TISSUE_LIST_PARENT here?

File: src/main/java/org/broadinstitute/dsm/route/participant/GetParticipantRoute.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+public class GetParticipantRoute extends RequestHandler {
+
+    @Override
+    protected Object processRequest(Request request, Response response, String userId) throws Exception {
+
+        QueryParamsMap queryParamsMap = request.queryMap();
+
+        String requestUserId = queryParamsMap.get(UserUtil.USER_ID).value();
+        if (!userId.equals(requestUserId)) throw new IllegalAccessException("User id: " + userId  + "does not match to request user id: " + requestUserId);
+
```
Comment (SimoneMaiwald) [inline src/main/java/org/broadinstitute/dsm/route/participant/GetParticipantRoute.java:27]: I think we should also check if the user is allowed to see pt list (same roles as in FilterRoute)

File: src/main/java/org/broadinstitute/dsm/model/filter/FilterFactory.java (hunk 1 above)
Comment (gmakharat) [reply src/main/java/org/broadinstitute/dsm/model/filter/FilterFactory.java:47]: Good question, I meant this when I mentioned poor design of filtering overall, frontend side is such organized that search bar filtering for tissue list parent comes this way, and search bar filtering for participant list parent comes totally different way.

File: src/main/java/org/broadinstitute/dsm/model/filter/participant/BaseFilterParticipantList.java (hunk 2 above)
Comment (gmakharat) [reply src/main/java/org/broadinstitute/dsm/model/filter/participant/BaseFilterParticipantList.java:81]: Can't say, main idea of refactoring is to keep the current behavior but change code structure to make it more organized, that's why I did not go deep details how logic is implemented, as I mentioned in my `P.S` filtering feature needs to be improved(logic) but not in refactoring part.

**Review by SimoneMaiwald — APPROVED** [review]

**Pull request merged**